Dockerize Laravel Stack

oliviasculley opened the pull request:
This PR attempts to dockerize the Laravel stack, and provides some documentation to hopefully initialize a reproducible docker setup from scratch.

Please let me know if you run into any issues trying to initialize the project through docker, or if you have any other comments. Thanks!

Closes #105

zach2825 commented:
Looks good, just a little feedback.

I resolved a conflict coming from the develop branch.

Check out [Laravel sail](https://laravel.com/docs/8.x/sail#starting-and-stopping-sail). laravel sail has a bunch of built-in helpers to use docker. Like `./vendor/bin/sail up` starts the docker, and `./vendor/bin/sail down` stop them. There are other useful commands.

You can also do things like the artisan commands, but with sail like `./vendor/bin/sail artisan migrate` for example
https://laravel.com/docs/8.x/sail#executing-sail-commands

oliviasculley commented:
Sail is built from docker pretty extensively, and is what I referenced heavily for this PR. https://tech.osteel.me/posts/you-dont-need-laravel-sail was super helpful to me in understanding what Sail was, but I think just using a stock docker setup uncovers a bit less of the magic, is more accessible, and will let people iterate and debug easier

zach2825 commented:
That's fair, to be honest I've never had luck using docker. I like what you
did here it loads up and servers the app just if you need to do artisan
commands sail makes that easier.

Thank you 😊

On Tue, Aug 31, 2021, 9:06 AM Olivia Sculley ***@***.***>
wrote:

> Sail is built from docker pretty extensively, and is what I referenced
> heavily for this PR.
> https://tech.osteel.me/posts/you-dont-need-laravel-sail was super helpful
> to me in understanding what Sail was, but I think just using a stock docker
> setup uncovers a bit less of the magic, is more accessible, and will let
> people iterate and debug easier
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/codeforgreenville/hackgreenville-com/pull/108#issuecomment-909218688>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA64TKK2QN6XUPTQ4XBBUKLT7THUJANCNFSM5DDG2BKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

zach2825 reviewed (approved):
Docker loads for me, I'm able to navigate the site in it. I'm not sure if anything else is needed but it's working

allella reviewed (commented):
My only question is if we need the .env.docker.  Is that a common pattern? If not, could we move the Docker specific configuration into .env.example and suggest people uncomment and/or adjust Docker config lines in there .env before running Docker.

csculley commented on CONTRIBUTING.md:
```diff
@@ -137,6 +137,24 @@ The HackGreenville project is now up and running! You should be able to open [lo
+The docker setup of this project should only be done for advanced users, or if
+needed for runtime compatibility issues.
+
+You need to make a copy of the `.env.docker` file and rename it to `.env` at the
```
I think I've included some env var changes to `.env.docker` that are docker specific, so I put them in `.env.docker` as not to mess with normal host installs. It's a common pattern from what I've seen, where people copy either `.env.example` or `.env.docker` to `.env` which is usually gitignored and contains secrets

oliviasculley replied:
I think I've included some env var changes to `.env.docker` that are docker specific, so I put them in `.env.docker` as not to mess with normal host installs. It's a common pattern from what I've seen, where people copy either `.env.example` or `.env.docker` to `.env` which is usually gitignored and contains secrets

allella commented:
Alright, I'm giving it a try and getting an error on _docker-compose pull_
Is this a temporary network error?

![image](https://user-images.githubusercontent.com/1777776/131703911-78fd44a9-5703-4936-82f4-2b0ac3f4f26e.png)

zach2825 commented:
I got that too, one time all 3 failed. re-try a few more times to see if it works for you

allella commented:
Still not working for me.

The docker-compose.yml has a [context](https://docs.docker.com/compose/compose-file/compose-file-v3/#context) of

`context: ./vendor/laravel/sail/runtimes/8.0`

but that directory doesn't exist if I clone the repo and continue with the Docker steps.

Does composer need to be run locally in order for the Sail image to be referenced in this way?

It seems Docker is trying to pull an image from a local directory and since it doesn't exist I'm guessing it's treating the context as a URL, which is likely why I'm getting the error

```
ERROR: for laravel.test  Get https://sail-8.0/v2/: dial tcp: lookup sail-8.0: no such host
ERROR: Get https://sail-8.0/v2/: dial tcp: lookup sail-8.0: no such host

```

> Either a path to a directory containing a Dockerfile, or a url to a git repository.
> 
> When the value supplied is a relative path, it is interpreted as relative to the location of the Compose file. This directory is also the build context that is sent to the Docker daemon.

oliviasculley commented:
I forgot to include instructions on how to fetch the Sail dependencies, so I've added instructions that utilize composer and alternatively just plain git to fetch the needed dependencies before building. I've also added additional instructions regarding the file permissions and build flag

allella commented:
My issues have been resolved after @oliviasculley 's latest commits.

allella commented:
Just noticed this was to the master branch.  We have been using the develop branch as the main branch for the stage website. Probably good for one of us to merge master back to develop to sync up. Or, perhaps there's a better branching strategy.

oliviasculley commented:
I can also open a PR to `develop` if you'd like!

allella commented:
Sure. We've been merging PRs into develop and then I'd do a release merge from develop to master after testing was done on the staging website.
